File: app/src/androidTest/java/com/github/jameshnsears/quoteunquote/DatabaseTestHelper.java
package com.github.jameshnsears.quoteunquote;

import com.github.jameshnsears.quoteunquote.database.DatabaseRepository;
import com.github.jameshnsears.quoteunquote.database.history.AbstractHistoryDatabase;
import com.github.jameshnsears.quoteunquote.database.quotation.QuotationEntity;
import com.github.jameshnsears.quoteunquote.utils.ContentType;

import org.junit.After;
import org.junit.Before;
import org.junit.Rule;

import java.util.ArrayList;
import java.util.List;

import androidx.arch.core.executor.testing.CountingTaskExecutorRule;
import androidx.arch.core.executor.testing.InstantTaskExecutorRule;
import androidx.test.core.app.ApplicationProvider;

import static org.junit.Assert.assertEquals;

public class DatabaseTestHelper {
    @Rule
    public InstantTaskExecutorRule instantTaskExecutorRule = new InstantTaskExecutorRule();

    @Rule
    public CountingTaskExecutorRule countingTaskExecutorRule = new CountingTaskExecutorRule();

    protected static int widgetID = 1;
    protected QuoteUnquoteModelFake quoteUnquoteModel;

    @Before
    public void setUp() {
        ApplicationProvider.getApplicationContext().deleteDatabase(AbstractHistoryDatabase.DATABASE_NAME);
        quoteUnquoteModel = new QuoteUnquoteModelFake();
    }

    @After
    public void tearDown() {
        quoteUnquoteModel.shutdown();
    }

    protected void insertTestDataSet01() {
        final List<QuotationEntity> quotationEntityList = new ArrayList<>();
        quotationEntityList.add(
                new QuotationEntity(
                        DatabaseRepository.DEFAULT_QUOTATION_DIGEST, "a0", "q0"
                ));
        quotationEntityList.add(
                new QuotationEntity(
                        "d1", "a1", "q1"
                ));
        quoteUnquoteModel.getDatabaseRepositoryFake().insertQuotations(quotationEntityList);
    }

    protected void insertTestDataSet02() {
        final List<QuotationEntity> quotationEntityList = new ArrayList<>();
        quotationEntityList.add(
                new QuotationEntity(
                        "d2", "a2", "q1"
                ));
        quotationEntityList.add(
                new QuotationEntity(
                        "d3", "a2", "q3"
                ));
        quotationEntityList.add(
                new QuotationEntity(
                        "d4", "a4", "q1"
                ));
        quoteUnquoteModel.getDatabaseRepositoryFake().insertQuotations(quotationEntityList);
    }

    protected void insertTestDataSet03() {
        final List<QuotationEntity> quotationEntityList = new ArrayList<>();
        quotationEntityList.add(
                new QuotationEntity(
                        "d5", "a5", "q1"
                ));
        quotationEntityList.add(
                new QuotationEntity(
                        "d6", "a2", "q6"
                ));
        quoteUnquoteModel.getDatabaseRepositoryFake().insertQuotations(quotationEntityList);
    }

    protected void setDefaultQuotation() {
        quoteUnquoteModel.setDefaultQuotation(widgetID);

        assertEquals(DatabaseRepository.DEFAULT_QUOTATION_DIGEST,
                quoteUnquoteModel.getNext(
                        widgetID, ContentType.ALL).digest);

        assertEquals(
                1,
                quoteUnquoteModel.countPrevious(widgetID, ContentType.ALL));
    }
}

File: app/src/androidTest/kotlin/com/github/jameshnsears/quoteunquote/db/DatabaseTestHelper.kt
package com.github.jameshnsears.quoteunquote.db

import android.content.Context
import android.os.Build
import androidx.annotation.NonNull
import androidx.test.core.app.ApplicationProvider
import com.github.jameshnsears.quoteunquote.db.q.QuotationEntity
import com.github.jameshnsears.quoteunquote.utils.ContentSelection
import com.github.jameshnsears.quoteunquote.utils.preference.PreferencesFacade
import org.hamcrest.CoreMatchers.equalTo
import org.hamcrest.MatcherAssert.assertThat
import org.junit.Before

abstract class DatabaseTestHelper {
    @NonNull
    var databaseRepositoryDouble: DatabaseRepositoryDouble =
        DatabaseRepositoryDouble.getInstance(ApplicationProvider.getApplicationContext())

    @JvmField
    val context: Context = ApplicationProvider.getApplicationContext()

    @Before
    open fun before() {
        databaseRepositoryDouble.eraseAllDatabsaes()
        PreferencesFacade.erase(context)
    }

    fun insertQuotationTestData01(useInternalDatabase: Boolean) {
        val quotationEntityList: MutableList<QuotationEntity> = ArrayList()
        quotationEntityList.add(
            QuotationEntity(
                DatabaseRepository.getDefaultQuotationDigest(useInternalDatabase),
                "w1",
                "a0",
                "q0",
            ),
        )
        quotationEntityList.add(QuotationEntity("d1234567", "w1", "a1", "q1"))
        databaseRepositoryDouble.insertQuotations(useInternalDatabase, quotationEntityList)
    }

    fun insertQuotationTestData02(useInternalDatabase: Boolean) {
        val quotationEntityList: MutableList<QuotationEntity> = ArrayList()
        quotationEntityList.add(QuotationEntity("d2345678", "w1", "a2", "q1"))
        quotationEntityList.add(QuotationEntity("d3456789", "w1", "a2", "q3"))
        quotationEntityList.add(QuotationEntity("d4567890", "w1", "a4", "q1"))
        databaseRepositoryDouble.insertQuotations(useInternalDatabase, quotationEntityList)
    }

    fun insertQuotationTestData03(useInternalDatabase: Boolean) {
        val quotationEntityList: MutableList<QuotationEntity> = ArrayList()
        quotationEntityList.add(QuotationEntity("d5678901", "w1", "a5", "q1"))
        quotationEntityList.add(QuotationEntity("d6789012", "w1", "a2", "q6"))
        databaseRepositoryDouble.insertQuotations(useInternalDatabase, quotationEntityList)
    }

    fun setDefaultQuotationAll(useInternalDatabase: Boolean, widgetId: Int) {
        databaseRepositoryDouble.markAsPrevious(
            useInternalDatabase,
            widgetId,
            ContentSelection.ALL,
            getDefaultQuotation(useInternalDatabase).digest,
        )

        databaseRepositoryDouble.markAsCurrent(
            useInternalDatabase,
            widgetId,
            getDefaultQuotation(useInternalDatabase).digest,
        )
    }

    fun setDefaultQuotationAuthor(useInternalDatabase: Boolean, widgetId: Int) {
        databaseRepositoryDouble.markAsPrevious(
            useInternalDatabase,
            widgetId,
            ContentSelection.AUTHOR,
            getDefaultQuotation(useInternalDatabase).digest,
        )
    }

    fun setDefaultQuotationSearch(useInternalDatabase: Boolean, widgetId: Int) {
        databaseRepositoryDouble.markAsPrevious(
            useInternalDatabase,
            widgetId,
            ContentSelection.SEARCH,
            getDefaultQuotation(useInternalDatabase).digest,
        )
    }

    fun markDefaultQuotationAsFavourite(useInternalDatabase: Boolean) {
        databaseRepositoryDouble.markAsFavourite(useInternalDatabase, getDefaultQuotation(useInternalDatabase).digest)
    }

    fun getDefaultQuotation(useInternalDatabase: Boolean): QuotationEntity =
        databaseRepositoryDouble.getQuotation(useInternalDatabase, DatabaseRepository.getDefaultQuotationDigest(useInternalDatabase))

    fun populateInternal(widgetId: Int) {
        insertInternalQuotations()

        databaseRepositoryDouble.markAsCurrent(
            true,
            widgetId,
            "d1234567",
        )

        databaseRepositoryDouble.markAsPrevious(
            true,
            widgetId,
            ContentSelection.ALL,
            "d1234567",
        )

        databaseRepositoryDouble.markAsFavourite(
            true,
            databaseRepositoryDouble.getCurrentQuotation(true, widgetId).digest,
        )

        assertThat(databaseRepositoryDouble.countAll(true).blockingGet(), equalTo(5))
    }

    fun insertInternalQuotations() {
        insertQuotationTestData01(true)
        insertQuotationTestData02(true)
    }

    fun populateExternal(widgetId: Int) {
        insertExternalQuotations()

        databaseRepositoryDouble.markAsCurrent(
            false,
            widgetId,
            DatabaseRepository.getDefaultQuotationDigest(false),
        )

        databaseRepositoryDouble.markAsPrevious(
            false,
            widgetId,
            ContentSelection.ALL,
            DatabaseRepository.getDefaultQuotationDigest(false),
        )

        databaseRepositoryDouble.markAsFavourite(
            false,
            databaseRepositoryDouble.getCurrentQuotation(false, widgetId).digest,
        )

        assertThat(databaseRepositoryDouble.countAll(false).blockingGet(), equalTo(2))
    }

    fun insertExternalQuotations() {
        val quotationEntityList: MutableList<QuotationEntity> = ArrayList()
        quotationEntityList.add(
            QuotationEntity(
                DatabaseRepository.getDefaultQuotationDigest(false),
                "",
                "external_a0",
                "external_q0",
            ),
        )
        quotationEntityList.add(
            QuotationEntity(
                "00000001",
                "",
                "external_a1",
                "external_q1",
            ),
        )
        databaseRepositoryDouble.insertQuotations(false, quotationEntityList)
    }

    protected fun canWorkWithMockk() = Build.VERSION.SDK_INT >= Build.VERSION_CODES.Q
}
